cmd/snap-discard-ns/snap-discard-ns.c

```diff
@@ -15,41 +15,190 @@
+#include <dirent.h>
 #include <errno.h>
+#include <fcntl.h>
+#include <fnmatch.h>
 #include <limits.h>
+#include <linux/magic.h>
+#include <stdio.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/vfs.h>
 #include <unistd.h>
 
 #include "../libsnap-confine-private/locking.h"
 #include "../libsnap-confine-private/string-utils.h"
 #include "../libsnap-confine-private/utils.h"
-#include "../snap-confine/ns-support.h"
+
+#ifndef NSFS_MAGIC
+#define NSFS_MAGIC 0x6e736673
+#endif
 
 int main(int argc, char **argv)
 {
-	if (argc != 2)
-		die("Usage: %s snap-name", argv[0]);
-	const char *snap_name = argv[1];
-
-	int snap_lock_fd = sc_lock_snap(snap_name);
-	debug("initializing mount namespace: %s", snap_name);
-	struct sc_mount_ns *group =
-	    sc_open_mount_ns(snap_name, SC_NS_FAIL_GRACEFULLY);
-	if (group != NULL) {
-		sc_discard_preserved_mount_ns(group);
-		sc_close_mount_ns(group);
+	if (argc != 2) {
+		printf("Usage: snap-discard-ns <SNAP-INSTANCE-NAME>\n");
+		return 0;
 	}
-	// Unlink the current mount profile, if any.
-	char profile_path[PATH_MAX] = { 0 };
-	sc_must_snprintf(profile_path, sizeof(profile_path),
-			 "/run/snapd/ns/snap.%s.fstab", snap_name);
-	if (unlink(profile_path) < 0) {
-		// Silently ignore ENOENT as the profile doens't have to be there.
-		if (errno != ENOENT) {
-			die("cannot remove current mount profile: %s",
-			    profile_path);
+
+	const char *snap_instance_name = argv[1];
+
+	/* Grab the lock holding the snap instance. This prevents races from
+	 * concurrently executing snap-confine. The lock is explicitly released
+	 * during normal operation but it is not preserved across the life-cycle of
+	 * the process anyway so no attempt is made to unlock it ahead of any call
+	 * to die() */
+	int snap_lock_fd = sc_lock_snap(snap_instance_name);
+	debug("discarding mount namespaces of snap %s", snap_instance_name);
+
+	const char *ns_dir_path = "/run/snapd/ns";
+	int ns_dir_fd = open(ns_dir_path, O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
+	if (ns_dir_fd < 0) {
+		/* The directory may legitimately not exist if no snap has started to
+		 * prepare it. This is not an error condition. */
+		if (errno == ENOENT) {
+			return 0;
 		}
+		die("cannot open path %s", ns_dir_path);
+	}
+
+	/* Move to the namespace directory. This is used so that we don't need to
+	 * traverse the path over and over in our upcoming unmount2(2) calls. */
+	if (fchdir(ns_dir_fd) < 0) {
+		die("cannot move to directory %s", ns_dir_path);
 	}
 
+	/* Create shell patterns that describe the things we are interested in:
+	 *
+	 * Preserved mount namespaces to unmount and unlink:
+	 * - "$SNAP_INSTANCE_NAME.mnt"
+	 * - "$SNAP_INSTANCE_NAME.[0-9]+.mnt"
+	 *
+	 * Applied mount profiles to unlink:
+	 * - "snap.$SNAP_INSTANCE_NAME.fstab"
+	 * - "snap.$SNAP_INSTANCE_NAME.[0-9]+.fstab" */
+	char sys_fstab_pattern[PATH_MAX];
+	char usr_fstab_pattern[PATH_MAX];
+	char sys_mnt_pattern[PATH_MAX];
+	char usr_mnt_pattern[PATH_MAX];
```

Comment:
Not a real problem in this case since the resulting buffer will be large enough for our short patterns, but the use of PATH_MAX for patterns is slightly confusing here.

Reply:
I will add a comment explaining that since those are path-related globs we use PATH_MAX to be on the safe side.